fix: mca display a perfromance warning for "sparse" dataset

```bash 

.venv/lib/python3.12/site-packages/lsg/projector/saiph.py:81: in transform
    transformed_df: pd.DataFrame = saiph.transform(model=self.model, df=df)
.venv/lib/python3.12/site-packages/saiph/projection.py:236: in transform
    return famd.transform(df, model)
.venv/lib/python3.12/site-packages/saiph/reduction/famd.py:292: in transform
    df_scaled = scaler(model, df)
.venv/lib/python3.12/site-packages/saiph/reduction/famd.py:268: in scaler
    df_quali[mod] = 0
.venv/lib/python3.12/site-packages/pandas/core/frame.py:4311: in __setitem__
    self._set_item(key, value)
.venv/lib/python3.12/site-packages/pandas/core/frame.py:4538: in _set_item
    self._set_item_mgr(key, value, refs)
.venv/lib/python3.12/site-packages/pandas/core/frame.py:4488: in _set_item_mgr
    self._mgr.insert(len(self._info_axis), key, value, refs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = BlockManager
Items: Index(['name___User_1', 'name___User_10', 'name___User_100', 'name___User_101',
       'name___Use...: int64
NumpyBlock: slice(1033, 1034, 1), 1 x 600, dtype: int64
NumpyBlock: slice(1034, 1035, 1), 1 x 600, dtype: int64
loc = 1034, item = 'name___User_348'
value = array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0]])
refs = None

    def insert(self, loc: int, item: Hashable, value: ArrayLike, refs=None) -> None:
        """
        Insert item at selected position.

        Parameters
        ----------
        loc : int
        item : hashable
        value : np.ndarray or ExtensionArray
        refs : The reference tracking object of the value to set.
        """
        with warnings.catch_warnings():
            # TODO: re-issue this with setitem-specific message?
            warnings.filterwarnings(
                "ignore",
                "The behavior of Index.insert with object-dtype is deprecated",
                category=FutureWarning,
            )
            new_axis = self.items.insert(loc, item)

        if value.ndim == 2:
            value = value.T
            if len(value) > 1:
                raise ValueError(
                    f"Expected a 1D array, got an array with shape {value.T.shape}"
                )
        else:
            value = ensure_block_shape(value, ndim=self.ndim)

        bp = BlockPlacement(slice(loc, loc + 1))
        block = new_block_2d(values=value, placement=bp, refs=refs)

        if not len(self.blocks):
            # Fastpath
            self._blklocs = np.array([0], dtype=np.intp)
            self._blknos = np.array([0], dtype=np.intp)
        else:
            self._insert_update_mgr_locs(loc)
            self._insert_update_blklocs_and_blknos(loc)

        self.axes[0] = new_axis
        self.blocks += (block,)

        self._known_consolidated = False

        if sum(not block.is_extension for block in self.blocks) > 100:
>           warnings.warn(
                "DataFrame is highly fragmented.  This is usually the result "
                "of calling `frame.insert` many times, which has poor performance.  "
                "Consider joining all columns at once using pd.concat(axis=1) "
                "instead. To get a de-fragmented frame, use `newframe = frame.copy()`",
                PerformanceWarning,
                stacklevel=find_stack_level(),
            )
E           pandas.errors.PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`

```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -16,6 +16,7 @@ dependencies = [
     "toolz>=1",
     "pydantic>=2",
     "msgspec>=0.19",
+    "matplotlib>=3.10.9",
 ]
 
 [project.optional-dependencies]

--- a/saiph/reduction/famd.py
+++ b/saiph/reduction/famd.py
@@ -1,6 +1,7 @@
 """FAMD projection module."""
 
 import sys
+from collections import defaultdict
 from collections.abc import Callable
 from itertools import chain, repeat
 from typing import Any, cast
@@ -250,17 +251,30 @@ def scaler(model: Model, df: pd.DataFrame) -> pd.DataFrame:
     df_quanti = (df_quanti - model.mean) / std_without_zeros
 
     # scale
-    df_quali = pd.get_dummies(
-        df[model.original_categorical].astype("category"),
-        prefix_sep=DUMMIES_SEPARATOR,
-        dtype=np.uint8,
-    )
-    # Here we add a column with 0 if the modality is not present in the dataset but
-    # was used to train the saiph model
+    # Attach fit-time categories to each column so that pd.get_dummies emits a
+    # column for every modality seen during fit — including those absent from
+    # this transform batch (they become all-zero columns automatically).
+    # This replaces a one-by-one loop that fragmented the DataFrame block
+    # manager and emitted a PerformanceWarning after 100+ insertions.
     if model._modalities is not None:
+        col_cats: dict[str, list[str]] = defaultdict(list)
         for mod in model._modalities:
-            if mod not in df_quali:
-                df_quali[mod] = 0
+            col, _, val = mod.partition(DUMMIES_SEPARATOR)
+            col_cats[col].append(val)
+
+        df_quali = pd.get_dummies(
+            df[model.original_categorical].assign(
+                **{col: pd.Categorical(df[col], categories=cats) for col, cats in col_cats.items()}
+            ),
+            prefix_sep=DUMMIES_SEPARATOR,
+            dtype=np.uint8,
+        )
+    else:
+        df_quali = pd.get_dummies(
+            df[model.original_categorical].astype("category"),
+            prefix_sep=DUMMIES_SEPARATOR,
+            dtype=np.uint8,
+        )
     df_quali = df_quali[model._modalities]
     df_quali = (df_quali - model.prop) / np.sqrt(model.prop)
 

--- a/saiph/reduction/famd_test.py
+++ b/saiph/reduction/famd_test.py
@@ -290,3 +290,111 @@ def test_reconstructed_df_from_weighted_model_equals_df() -> None:
     model = fit(df, col_weights=[3, 1, 1, 1, 1])  # type: ignore
     reconstructed_df = reconstruct_df_from_model(model)
     assert_frame_equal(df, reconstructed_df)
+
+
+# ---------------------------------------------------------------------------
+# Absent / novel modality tests
+# ---------------------------------------------------------------------------
+# These three tests collectively protect the behaviour of scaler() under the
+# conditions that actually occur in the avatar pipeline:
+#
+#   • Privacy metrics fit on 50 % of rows, then transform the other 50 %
+#     (→ holdout may be missing categories seen only in the training half).
+#   • Avatar data is transformed with the original model after generation
+#     (→ rare categories may be absent from the avatar batch).
+#   • Cross-table linkage transforms a child table with a parent-table model
+#     (→ child may have categories never seen in the parent).
+# ---------------------------------------------------------------------------
+
+
+def test_transform_famd_absent_categories_no_performance_warning() -> None:
+    """FAMD (mixed continuous + categorical) must not emit PerformanceWarning
+    when >100 categories are absent from the transform batch.
+
+    The existing MCA-only test covers pure-categorical input (mca.py).
+    This test covers the mixed-type path through famd.scaler, which is the
+    real-world call site in the avatar quality-metrics pipeline.
+
+    pytest is configured to turn all warnings into errors, so this test
+    currently *fails* with the fragmentation loop and will pass once the
+    loop is replaced by a single vectorised operation.
+    """
+    rng = np.random.default_rng(42)
+    n_categories = 150  # exceeds the pandas >100 block-fragmentation threshold
+    n_rows = n_categories * 4
+
+    df_train = pd.DataFrame(
+        {
+            "num": rng.normal(size=n_rows),
+            "cat": [f"val_{i}" for i in range(n_categories)] * 4,
+        }
+    )
+    _, model = fit_transform(df_train, nf=5)
+
+    # Only the first 50 categories are present → 100 of 150 are absent.
+    df_test = pd.DataFrame(
+        {
+            "num": rng.normal(size=50),
+            "cat": [f"val_{i}" for i in range(50)],
+        }
+    )
+    coord_test = transform(df_test, model)
+
+    assert coord_test.shape == (50, 5)
+
+
+def test_transform_coordinates_independent_of_batch_composition() -> None:
+    """Coordinates for a row must not change depending on which other categories
+    happen to appear in the same transform batch.
+
+    This verifies the mathematical invariant: absent categories are filled with
+    zero *before* centering and scaling, using the fitted model.prop values.
+    The result for a given row is therefore a pure function of that row’s own
+    values plus the fitted model, with no dependency on the batch.
+    """
+    rng = np.random.default_rng(0)
+    n = 30
+    df_train = pd.DataFrame(
+        {
+            "num": rng.normal(size=n),
+            "cat": rng.choice(["a", "b", "c"], n),
+        }
+    )
+    _, model = fit_transform(df_train)
+
+    # Row under test: num=1.0, cat="a"
+    # Batch A: only category "a" present — "b" and "c" are absent.
+    coords_isolated = transform(pd.DataFrame({"num": [1.0], "cat": ["a"]}), model)
+
+    # Batch B: all three categories present — the target row is still num=1.0 / cat="a".
+    coords_mixed_batch = transform(
+        pd.DataFrame({"num": [1.0, 2.0, 3.0], "cat": ["a", "b", "c"]}), model
+    ).iloc[[0]]
+
+    assert_frame_equal(
+        coords_isolated.reset_index(drop=True),
+        coords_mixed_batch.reset_index(drop=True),
+    )
+
+
+def test_transform_novel_category_not_seen_at_fit() -> None:
+    """A category value that was not present during fit must be silently ignored.
+
+    Rows containing an unknown category are treated as if all model-known dummy
+    columns for that variable are zero — i.e., the same as an absent known
+    category.  The transform must complete without error and return finite,
+    non-NaN coordinates.
+    """
+    df_train = pd.DataFrame(
+        {
+            "num": [1.0, 2.0, 3.0, 4.0],
+            "cat": ["a", "a", "b", "b"],
+        }
+    )
+    _, model = fit_transform(df_train)
+
+    df_novel = pd.DataFrame({"num": [1.5], "cat": ["unseen_value"]})
+    coord = transform(df_novel, model)
+
+    assert coord.shape == (1, model.nf)
+    assert not coord.isnull().any().any()

